app/components/MazemapEmbed/index.js

```diff
@@ -0,0 +1,82 @@
+// @flow
+import { useEffect } from 'react';
+import 'node_modules/mazemap/mazemap.min.css';
+import * as Mazemap from 'mazemap';
+import MazemapLink from './MazemapLink';
+
+type Props = {
+  mazemapPoi: number,
+  height?: number,
+  linkText?: string,
+};
+
+/** A component that shows a mazemap map of a given poi (e.g. room),
+ * largely based on https://api.mazemap.com/js/v2.0.12/docs/#ex-data-poi
+ */
+
+export const MazemapEmbed = ({ mazemapPoi, ...props }: Props) => {
+  //useEffect to initialize only once, mazemapPoi will probably not change
+  useEffect(() => {
+    const embeddedMazemap = new Mazemap.Map({
+      container: 'mazemap-embed',
+      campuses: 1,
+      center: { lng: 10.4042965, lat: 63.4154135 },
+      zLevel: 3,
+      zoom: 16,
+      minZoom: 10,
+      maxZoom: 20,
+      zLevelControl: true,
+      scrollZoom: true,
+      doubleClickZoom: false,
+      touchZoomRotate: false,
+    });
+    embeddedMazemap.on('load', () => {
+      // Initialize a Highlighter for POIs
+      // Storing the object on the map just makes it easy to access for other things
+      embeddedMazemap.highlighter = new Mazemap.Highlighter(embeddedMazemap, {
+        showOutline: true,
+        showFill: true,
+        outlineColor: Mazemap.Util.Colors.MazeColors.MazeRed,
+        fillColor: Mazemap.Util.Colors.MazeColors.MazeRed,
+      });
+      // Fetching via Data API
+      Mazemap.Data.getPoi(mazemapPoi).then((poi) => {
+        placePoiMarker(poi);
+      });
+
+      const placePoiMarker = (poi) => {
+        // Get a center point for the POI, because the data can return a polygon instead of just a point sometimes
+        const lngLat = Mazemap.Util.getPoiLngLat(poi);
+
+        new Mazemap.MazeMarker({
+          color: '#c0392b',
+          innerCircle: true,
+          innerCircleColor: '#fff',
+          size: 34,
+          innerCircleScale: 0.5,
+          zLevel: poi.properties.zLevel,
+        })
+          .setLngLat(lngLat)
+          .addTo(embeddedMazemap);
+
+        // If we have a polygon, use the default 'highlight' function to draw a marked outline around the POI.
+        if (poi.geometry.type === 'Polygon') {
+          embeddedMazemap.highlighter.highlight(poi);
+        }
+        embeddedMazemap.jumpTo({ center: lngLat, zoom: 18 });
+        embeddedMazemap.zLevel = poi.properties.zLevel;
+      };
+
+      const height = embeddedMazemap.getCanvas().clientHeight;
+      const maxHeight = height - 50; // 50 pixels account for margins and spacing
+      embeddedMazemap.zLevelControl.setMaxHeight(maxHeight);
+    });
+  }, [mazemapPoi]);
+
+  return (
+    <>
+      <div style={{ height: props.height || 400 }} id="mazemap-embed" />
+      <MazemapLink mazemapPoi={mazemapPoi} linkText={props.linkText} />
+    </>
+  );
+};
```

Comment:
Is this tested with the server-side renderer? I have a slight suspicion it might break.

It can be run with `yarn:ssr` and if there are any errors in the console or/and you can see the page loading (some js doing something) it's broken.

I don't have time to test this myself, but I'm sure someone else (@webkom/aktive) can do this real quick and make sure it's a-okay.  If not, we need to do some changes to make sure it doesn't break on the SSR (conditionally loading on the client f.ex.)

Reply:
I think your suspicions were correct. Map does load, but the console dislikes mazemap.min.js :/
```js
loadable-components: failed to synchronously load component, which expected to be available {  
  fileName: 85469,
  chunkName: 'meetings',
  error: 'Element is not defined'
}
08:15:38.559Z ERROR lego-webapp: render_error
    ReferenceError: Element is not defined  
        at eval (webpack://Mazemap/./src/Util/Polyfills.js?:21:5)
        at Object.5516 (/home/debian/lego-webapp/node_modules/mazemap/mazemap.min.js:1:99125)
        at __webpack_require__ (/home/debian/lego-webapp/node_modules/mazemap/mazemap.min.js:1:1911363)
        at eval (webpack://Mazemap/./src/mazemap.js_+_124_modules?:412:17)
        at Module.2061 (/home/debian/lego-webapp/node_modules/mazemap/mazemap.min.js:1:104806)
        at __webpack_require__ (/home/debian/lego-webapp/node_modules/mazemap/mazemap.min.js:1:1911363)
        at /home/debian/lego-webapp/node_modules/mazemap/mazemap.min.js:1:1911965
        at /home/debian/lego-webapp/node_modules/mazemap/mazemap.min.js:1:1912019
        at /home/debian/lego-webapp/node_modules/mazemap/mazemap.min.js:1:81
        at Object.<anonymous> (/home/debian/lego-webapp/node_modules/mazemap/mazemap.min.js:1:195)
```

Reply:
Yeah, that's what I feared. I think the best way to do this, is to not render the mazemap component on first render, for then to conditionally render it on the client after hydration. I.e. Allocate some "placeholder" wrapper where the map should be and only render that (so that the page content does not jump around, with the exact dimensions before the map is loaded).

You can then, on the client, render the component after the app is hydrated and the wrapper component is mounted. The proper way to do this is simply with a `useEffect` hook in the wrapper (since these don't run on render, only on mount). [This guide](https://tech.willhaben.at/how-to-shoot-yourself-in-the-foot-in-react-401e7dbba720)[^1] covers it pretty good how you should do this.

[^1]: He also links to [this article](https://www.joshwcomeau.com/react/the-perils-of-rehydration/) in the guide, which goes more in-depth into the problem if you are interested

Reply:
So I've fixed it for events by dynamically importing mazemap (seems the offending code is in the mazemap module), but meetings still have the issue for some reason.

Lazyloading seemed like a solution (```const RawMazemap = lazy(() => import('./RawMazemap'));```) where `RawMazemap` is the real mazemap component, and we wrap it in MazemapEmbed.  However, `Suspense` is experimental or not supported with ReactDOM, so I discarded that. 

I don't understand why only meetings get this error now, it seems to load `MazemapEmbed` in the exact same way as events. Maybe a wrapper that dynamically imports another component, which imports `mazemap` works, or maybe a small change to meetings fixes it?

Reply:
Lazyloading won't help here. Since the problem is in the mazemap module using DOM/browser APIs that are not available in Node.js. So the only way to fix the issue is as I stated in the comment and not rendering it on the server at all. 

Nice work with events, although I don't see why you need to dynamically import it? Unless the mazemap module contains code that runs on require? (which it really should not)

If that is the case, you could try lazy loading I guess. As you stated, native lazy loading does not support SSR yet, so we use react-loadable for this purpose. See the documentation and dynamic imports elsewhere in the app for usage.

Reply:
I tried using require('mazemap'), but then the element not defined error happens with both events and meetings, so it would seem mazemap runs some code on import.

Reply:
Right. But did you also just use es `import` as before? Importing es modules should not run any code, in contrast to cjs `require` statements.

Reply:
I used `import * as Mazemap from 'mazemap'` before, which gave the same error.

Reply:
Hmm that's strange. Did you remember to run `yarn build` and restart the SSR+server?

Reply:
I rebuilt and restarted everything on the frontend.

Reply:
Hmmm. Then it might run some code on import then. I would imagine the code you have with dynamic imports works then. So very strange of it doesn't work only on Meetings.

Reply:
Exactly, I cannot understand why it fails only with meetings when events use the exact same component.

Reply:
Soo, I just tested this with the latest from your branch, and both events and meetings seem to work just fine? :smiling_face_with_tear: 

Note that sometimes just after you start up the server, you might need to hard refresh a couple times before it's completely stable (due to caching and some weird states and all that). But anyways, this looks really good now. And afaikt this is good to go!

Reply:
Riiight! @ollfkaih I found the problem. The issue is not in the mazemapEmbed component. That works great. But the same issue occurs in the `mazemapAutocomplete` component of course. As that loads the same library. Why it works in some places and not others probably has to do with the way `reactDOM`, `webpack` and `node` traverses the dependency tree when rendering the app.

So just do the same thing in the `mazemapAutocomplete` component and then everything should work just dandy!